src/deepsparse/transformers/engines/nl_decoder_engine.py

```diff
@@ -0,0 +1,323 @@
+# Copyright (c) 2021 - present / Neuralmagic, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+from typing import Any, Dict, List, Optional, Tuple
+
+import numpy
+import onnx
+from transformers import AutoTokenizer
+
+from deepsparse.engine import Context
+from deepsparse.pipeline import create_engine
+from deepsparse.transformers.utils.decoder_kv_cache import DecoderKVCache
+from deepsparse.transformers.utils.helpers import generate_session_id, softmax
+from sparsezoo.utils.onnx import save_onnx
+
+
+_LOGGER = logging.getLogger(__name__)
+
+__all__ = ["NLDecoderEngine"]
+
+_CACHE_INPUT_NAME = "past_key_values"
+
+
+class NLDecoderEngine:
+    """
+    The NLDecoderEngine (NaturalLanguageDecoderEngine) handles the
+    logic around the inference for Natural Language pipeline,
+    including batching and kv cache logic.
+
+    :param onnx_file_path: The path to the onnx model file
+    :param engine_type: The type of engine to use for the inference
+    :param engine_args: The arguments to pass to the engine
+    :param sequence_length: The maximum sequence length to run the engine for
+    :param input_ids_length: The maximum input ids length to run the engine for
+    :param engine_context: The context to run the engine in
+    :param sampling_temperature: The temperature to use for sampling
+    :param deterministic: Whether to use deterministic sampling
+    :param tokenizer: The tokenizer to used for engine inputs
+    :param engine_context: The context to run the engine in
+    :param use_deepsparse_cache: Whether to use the deepsparse
+        kv cache in the DecoderKVCache object or not
+    """
+
+    def __init__(
+        self,
+        onnx_file_path: str,
+        engine_type: str,
+        engine_args: Dict[str, Any],
+        sequence_length: int,
+        input_ids_length: int,
+        tokenizer: AutoTokenizer,
+        sampling_temperature: float = 1.0,
+        deterministic: bool = True,
+        engine_context: Optional[Context] = None,
+        use_deepsparse_cache=False,
+    ):
+
+        onnx_file_path, kv_cache_enabled = self.overwrite_onnx_model_inputs(
+            onnx_file_path=onnx_file_path,
+            batch_size=engine_args.get("batch_size", 1),
+            sequence_length=sequence_length,
+            input_ids_length=input_ids_length,
+        )
+
+        self.engine = create_engine(
```

Comment:
don't we still want `cache_bools` propagated to the engine compilation when cache is enabled?

Reply:
Yes, this is definitely needed, thanks for pointing out.